src/lib/SaferNames/NameCore.hs

```diff
@@ -0,0 +1,294 @@
+-- the scope `[z, y]`. To prove that we can do the injection as an coercion,
+-- it's sufficient to know (1) that the new scope contains distinct names, and
+-- (2) it extends the old scope. These are the `Distinct l` and
+-- `(InjectableB b, b)` conditions below.
+
+injectNames :: InjectableE e => InjectableB b
+            => Distinct l => b n l -> e n -> e l
+injectNames _ x = unsafeCoerceE x
+
+-- `injectNames` is the only function we actualy use in practice. The rest of
+-- this section is just about proving that it's valid to use `injectNames`, by
+-- offering tools to implement `Injectable` instances.
+
+-- To prove that a data type is injectable by coercion, we have to show that
+-- it's injectable without using coercion. `injectionProof` is never actually
+-- called at run time, but it should still be implemented to prove that it's
+-- possible. This isn't watertight. For example, you could write
+-- `injectionProof = injectNames` and it'd still type check!
+class InjectableE (e::E) where
+  injectionProofE :: ObservablyFresh n l -> e n -> e l
+
+class InjectableB (b::B) where
+  injectionProofB :: ObservablyFresh n n' -> b n l
+                  -> (forall l'. ObservablyFresh l l' -> b n' l' -> a)
+                  -> a
+
+  -- Not sure if this belongs here, but it doesn't make sense in `HasNamesB`
+  -- either. Maybe we need a separate type class?
+  boundNames :: b n l -> NameSet (n:=>:l)
+
+-- an `ObservablyFresh n l` means that scope `l` is a superset of scope `n` that
+-- doesn't shadow any of the names exposed by `n`.
+data ObservablyFresh (n::S) (l::S) = UnsafeMakeObservablyFresh
+
+instance InjectableE (Name s) where
+  injectionProofE _ name = unsafeCoerceE name
+
+-- Here's the reasoning informally. Let's say `x` is the name of the binder.
+-- The scopes are related like this:
+--   l  = n  ++ [x]
+--   l' = n' ++ [x]
+-- if n' is observably fresh wrt n then l' is observably fresh wrt l. This is because every
+-- exposed name in l' is either
+--   (1) x itself, which is exposed in both l and l', thus l' doesn't shadow it, or
+--   (2) a name other than x exposed by n', which we know is observably fresh
+--       wrt n and thus also observably fresh wrt n++[x]
+instance InjectableB (NameBinder s) where
+  injectionProofB  _ b cont = cont UnsafeMakeObservablyFresh $ unsafeCoerceB b
+  boundNames (UnsafeMakeBinder name) = UnsafeMakeNameSet (S.singleton name)
+
+instance (forall s. InjectableE s => InjectableE (v s)) => InjectableE (NameMap v i) where
+
+-- === environments ===
+
+-- The `NameMap` type is purely an optimization. We could do everything using
+-- the safe API by defining:
+--    type NameMap v i o = (NameSet i, forall s. Name s i -> v s o)
+-- Instead, we use this unsafely-implemented data type for efficiency, to avoid
+-- long chains of case analyses as we extend environments one name at a time.
+
+data NameMap
+  (v::E -> E)  -- env payload, as a function of the static data type
```

Comment:
Oh, by "payload" I just mean "the thing you get when you look up a name in the env". In the case of a substitution, it's an `Atom`. Or it might be a name. Or a type. What's new is that the type of the thing you get when you query an env can now depend on the type parameter associated with the name. So you can have different sorts of names, like data definition names and let/lambda-bound names, that give you different things when you look them up in an env.

Reply:
It's really to handle two important sorts of envs: static info envs, and substitution envs. The static info envs just use the identity function on `E`, giving a payload that's exactly the type specified by the `s` parameter in `Name s n`. The substitution envs use `SubstVal`, which allow them to apply a substitution only to names with the correct `s` parameter, and a renaming otherwise.